### PR TITLE
move relation and turn relation parsing out of FlagEncoder

### PR DESCRIPTION
**Pull request by karussell**
This pull request removes several methods of FlagEncoders and TurnCostEncoder.java and is one of the many tasks to remove this class like #1625. Especially in FlagEncoders that do not use turn costs like FootFlagEncoder you see the effects of this change. Also OSMReader.TurnCostTableEntry is no longer part of the public API. 

Initially I wanted to create two PRs, one for the 'pure' relation and one for the turn restriction part, but I think this belongs together. 

So this PR is based on the branch: https://github.com/graphhopper/graphhopper/commits/move_rel_out_of_flag_encoder where the first two commits move the relation handling out of FlagEncoder into a new RelationTagParser (plus some clean up commits). It also fixes a bug here that calling the super method here has no side effects: https://github.com/graphhopper/graphhopper/blob/master/core/src/main/java/com/graphhopper/routing/util/MountainBikeFlagEncoder.java#L150 Additionally there was code reduction possible in some FlagEncoders. 

From commit 22b1483 it cleans up the turn relation parsing and moves it into OSMTurnRelationParser. The OSMReader shows its age here and needs some later rewrite maybe using mapdb or similar so that several workarounds can be avoided (this is tricky as it needs to be very fast). At the moment there are 2 phases: preprocessing and processing that both have 3 subprocedures: 

 * 1.1. preprocess nodes, 1.2. preprocess way, 1.3. preprocess relations
 * 2.1. handle nodes, 2.2. handle way, 2.3. handle relations

The RelationTagParser fetches the ReaderRelation tags in step 1.3 and compacts them into relationFlags (IntsRef) to convert them in step 2.2 into edgeFlags: see the new EncodedValues for RouteNetwork (`bike.network` and `foot.network`). These EncodedValues can be used in other TagParsers.

The TurnCostParser is more involved and different to the RelationTagParser as it needs a traversable Graph and e.g. for `car.turn_cost` it requires a ready-to-use `car.access`. So this work is done at step 2.3. Also the method TurnCostParser.handleTurnRelationTags requires several parameters like the graph and an interface to convert between OSM and internal GraphHopper IDs.

TODOs:

 - [x] cache EdgeExplorer creation in OSMTurnRelationParser like it was before?
 - [x] Should we use one parser instead of the two OSMFootNetworkTagParser and OSMBikeNetworkTagParser that have only tiny differences? -> keep it separate
 - [x] Make turn cost handling easier via https://github.com/graphhopper/graphhopper/issues/1772#issuecomment-547626445 and remove TurnCostAccess
 - [x] rename TurnCostExtension to TurnCostStorage?
 - [x] remove TODO NOWs

### CODE REVIEW

**Comment by easbar** [comment]
Copied from now closed #1772:

> Update: this would be also helpful for the branch move_tc_out_of_flag_encoder (#1775) and we could try if the following is possible: 

>   instead of TurnCostAccess we make usage simple via:
    edgeIteratorState.setTurnCost(carTurnCostEnc, nextEdgeId, 10.0) // base->adj->nextEdgeId
    we move tcFlags inside EdgeIteratorState (remove it e.g. in TurnWeighting and TurnCostAccess)
    we should be able to remove entire TurnCostAccess

File: core/src/test/java/com/graphhopper/routing/ch/CHTurnCostTest.java [hunk 1]
```diff
@@ -181,14 +181,14 @@ public void testFindPath_multipleInOutEdges_turnReplacementDifference() {
...
         setRandomCost(6, 7, 10, rnd);
-        setRandomCost(0, 5, 3, rnd);
-        setRandomCost(1, 5, 3, rnd);
-        setRandomCost(0, 5, 6, rnd);
-        setRandomCost(1, 5, 6, rnd);
-        setRandomCost(4, 7, 8, rnd);
-        setRandomCost(4, 7, 9, rnd);
-        setRandomCost(6, 7, 8, rnd);
-        setRandomCost(6, 7, 9, rnd);
+        setRandomCostOrRestriction(0, 5, 3, rnd);
```
Comment (easbar) [inline core/src/test/java/com/graphhopper/routing/ch/CHTurnCostTest.java]: just using setRandomCost would not lead to strict restrictions

File: core/src/test/java/com/graphhopper/storage/ShortcutUnpackerTest.java [hunk 2]
```diff
@@ -231,22 +231,21 @@ public void withTurnWeighting() {
...
+        setTurnCost(edge5.getEdge(), 6, NEXT_EDGE, 6.0);
         // turn costs <-
-        tce.setExpensive(encoder.toString(), encodingManager, NEXT_EDGE, 6, edge5.getEdge(), 2.0);
-        tce.setExpensive(encoder.toString(), encodingManager, edge5.getEdge(), edge4.getEdge(), 5, 3.0);
-        tce.setExpensive(encoder.toString(), encodingManager, edge4.getEdge(), edge3.getEdge(), 4, 2.0);
-        tce.setExpensive(encoder.toString(), encodingManager, edge3.getEdge(), edge2.getEdge(), 3, 4.0);
-        tce.setExpensive(encoder.toString(), encodingManager, edge2.getEdge(), edge1.getEdge(), 2, 1.0);
-        tce.setExpensive(encoder.toString(), encodingManager, edge1.getEdge(), edge0.getEdge(), 1, 0.0);
-        tce.setExpensive(encoder.toString(), encodingManager, edge0.getEdge(), 0, PREV_EDGE, 1.0);
+        setTurnCost(NEXT_EDGE, 6, edge5.getEdge(), 2.0);
```
Comment (easbar) [inline core/src/test/java/com/graphhopper/storage/ShortcutUnpackerTest.java]: here the via node and to edge were intermixed

File: core/src/test/java/com/graphhopper/routing/DirectedBidirectionalDijkstraTest.java [hunk 3]
```diff
@@ -250,11 +250,11 @@ public void directedRouting() {
         int leftNorth = graph.edge(9, 0, 1, true).getEdge();
 
         // make paths fully deterministic by applying some turn costs at junction node 2
-        addTurnCost(7, 2, 3, 1);
-        addTurnCost(7, 2, 6, 3);
-        addTurnCost(1, 2, 3, 5);
-        addTurnCost(1, 2, 6, 7);
-        addTurnCost(1, 2, 7, 9);
+        setTurnCost(7, 2, 3, 1);
```
Comment (easbar) [inline core/src/test/java/com/graphhopper/routing/DirectedBidirectionalDijkstraTest.java]: I think its better to do such renamings in a separate commit in master (same with turn cost extension/storage). I changed this in master and merged here again

File: core/src/main/java/com/graphhopper/routing/weighting/TurnWeighting.java [hunk 4]
```diff
@@ -108,22 +111,14 @@ public double calcTurnWeight(int edgeFrom, int nodeVia, int edgeTo) {
         if (!EdgeIterator.Edge.isValid(edgeFrom) || !EdgeIterator.Edge.isValid(edgeTo)) {
             return 0;
         }
+        double tCost = 0;
         if (turnCostExt.isUTurn(edgeFrom, edgeTo)) {
-            // note that the u-turn costs overwrite any turn costs set in TurnCostStorage
-            // todo:
-            // also this does not allow TurnCostEncoder to set the u-turn costs to zero explicitly, like FootFlagEncoder!
-            // this problem is a bit hidden, because if you do not apply any turn restrictions (like FootFlagEncoder)
-            // you never do a u-turn anyway.
```
Comment (easbar) [inline core/src/main/java/com/graphhopper/routing/weighting/TurnWeighting.java]: the comment still applies, no ? setting u-turn costs on turnCostExt has no effect, because the uTurnCosts take a higher priority here.

File: core/src/test/java/com/graphhopper/routing/ch/CHTurnCostTest.java (hunk 1 above)
Comment (karussell) [reply core/src/test/java/com/graphhopper/routing/ch/CHTurnCostTest.java]: Ups, thanks for the fix!

File: core/src/test/java/com/graphhopper/routing/DirectedBidirectionalDijkstraTest.java (hunk 3 above)
Comment (karussell) [reply core/src/test/java/com/graphhopper/routing/DirectedBidirectionalDijkstraTest.java]: Yes, you are right. Thanks!

File: core/src/main/java/com/graphhopper/routing/weighting/TurnWeighting.java (hunk 4 above)
Comment (karussell) [reply core/src/main/java/com/graphhopper/routing/weighting/TurnWeighting.java]: Ah, ok. This is what was meant with the comment before the todo. I'll revert the removal of that.

Regarding the TODO: the FootFlagEncoder does not have any turn restriction stuff but still overwriting/overruling is not possible. 

BTW: could we make isUTurn dependent on turnCostEnc and then remove this comment?

File: core/src/main/java/com/graphhopper/routing/profiles/TurnCost.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    /**
+     * You need to call EncodingManager.getKey(prefix, EV_SUFFIX) as this EncodedValue can be used for e.g. car and bike
+     */
+    public final static String EV_SUFFIX = "turn_cost";
+
+    /**
+     * This creates an EncodedValue specifically for the turn costs
+     */
+    public static DecimalEncodedValue create(String name, int maxTurnCosts) {
+        int turnBits = Helper.countBitValue(maxTurnCosts);
```
Comment (easbar) [inline core/src/main/java/com/graphhopper/routing/profiles/TurnCost.java]: Not sure if this is relevant here, but IntelliJ complains about `return counter++` in `Helper#countBitValue` (`counter is incremented "after" it is returned)

File: core/src/main/java/com/graphhopper/routing/util/parsers/OSMTurnRelationParser.java [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+            while (iter.next()) {
+                int edgeId = iter.getEdge();
+                long wayId = map.getOsmIdOfInternalEdge(edgeId);
+                if (edgeId != edgeIdFrom && osmTurnRelation.getRestriction() == OSMTurnRelation.Type.ONLY && wayId != osmTurnRelation.getOsmIdTo()
+                        || osmTurnRelation.getRestriction() == OSMTurnRelation.Type.NOT && wayId == osmTurnRelation.getOsmIdTo() && wayId >= 0) {
+                    final TCEntry entry = new TCEntry(edgeIdFrom, viaNode, iter.getEdge());
+                    IntsRef tcFlags = new IntsRef(turnCostFlags.length);
+                    getTurnCostEnc().setDecimal(false, tcFlags, Double.POSITIVE_INFINITY);
+                    entries.add(entry);
+                    tcs.setTurnCost(tcFlags, edgeIdFrom, viaNode, iter.getEdge());
```
Comment (easbar) [inline core/src/main/java/com/graphhopper/routing/util/parsers/OSMTurnRelationParser.java]: Why can we not use tcs.set(...) here and why do we (need to) create a new IntsRef everytime we are here ?

**Comment by easbar** [comment]
LGTM

File: core/src/main/java/com/graphhopper/routing/util/parsers/OSMTurnRelationParser.java (hunk 6 above)
Comment (karussell) [reply core/src/main/java/com/graphhopper/routing/util/parsers/OSMTurnRelationParser.java]: Ah, yes. It is indeed no longer required to create a new IntsRef now that we do no collect the TurnCostEntry. And will also try to use tcs.set...

**Comment by karussell** [comment]
Thanks a lot for the review - I finally could get rid of a bit more code :)
